### PR TITLE
Eio bikereg example

### PR DESCRIPTION
**Pull request by ada2k**
Hi, I've done a very simple translation of bikereg into Eio. Also made some minor tweaks to the readme's prerequisites section, and pinned ppx_rapper to allow caqti-eio to install (which it seems like was needed anyway).

### CODE REVIEW

**Comment by ada2k** [comment]
Partially closes #15

File: study/bikereg-eio/bikereg.ml [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
+ * License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * and the LGPL-3.0 Linking Exception along with this library.  If not, see
+ * <http://www.gnu.org/licenses/> and <https://spdx.org>, respectively.
+ *)
```
Comment (paurkedal) [inline study/bikereg-eio/bikereg.ml]: I've committed the LWT bikereg example with the correct license header.  Can you amend and force push when you update, so that it does not look like you changed the license i the Git history?

File: dune-project [hunk 2]
```diff
@@ -0,0 +1 @@
+(lang dune 3.12)
```
Comment (paurkedal) [inline dune-project]: This should be moved down, since we currently make each example an independent dune project.

File: README.md [hunk 3]
```diff
@@ -63,14 +63,17 @@ We will also:
...
-# Or install packages manually, see: dev.opam
-opam install . --deps-only
-```
+- Postgresql development libraries
+
+  These can be found in `libpq-dev` on Debian-based systems, `libpq-devel` on Fedora and `postgresql-libs` on Arch
+
+- Caqti and ppx_rapper
+
+  These can be installed via `opam install . --deps-only`, and will require OCaml 5 for the Eio examples. If you're on a lower version, you'll have to install the required packages from dev.opam manually.
```
Comment (paurkedal) [inline README.md]: I don't think this belongs to the current PR?

File: dev.opam [hunk 4]
```diff
@@ -37,6 +37,9 @@ depends: [
   "caqti-driver-postgresql"
   "caqti-lwt"
   "caqti-async"
+  "async"
```
Comment (paurkedal) [inline dev.opam]: Does not belong to this commit, though yes it might be a good idea to make all direct dependencies explicit.

File: dev.opam [hunk 5]
```diff
@@ -50,3 +53,7 @@ depends: [
   # Below is for documentation purposes
   "mdx"
 ]
+
+pin-depends: [
+  "ppx_rapper.3.1.0" "git+https://github.com/roddyyaga/ppx_rapper#2222edbbe68db7ba1ab0c7a2688c227ea5c0f230"
+]
```
Comment (paurkedal) [inline dev.opam]: Also, not in this commit, at least.

**Review by paurkedal — CHANGES_REQUESTED** [review]
I think this looks good if you remove the unrelated changes update the license header, move down the dune-project and rebase every on main (so that it looks like it was based on the in-repo bikereg example from the Git history).

**Comment by ada2k** [comment]
Removed the unrelated changes. Would it be useful to submit a PR for at least the opam file changes.

**Comment by benjamin-thomas** [comment]
Nice!

Indeed I forgot to add a few dependencies to the opam file, sorry about that. At a later point, I intend to write setup scripts so that we can minimize things going out of sync.

--

I have a couple of suggestions. Firstly, we have a nice mdx setup already, so by providing a `README.md` we could assert the output of this simple program by execute a shell script. 

So we could still ensure that everything subjproject works nicely by running this single command at the root of the project

```
PGHOST=localhost PGDATABASE=caqti_study PGPORT=5433 dune runtest -j1
```

For this to work, a little change would be required (my second suggestion): reporting the stolen bikes at the beginning of the year (this is to prevent mdx continously failing because the current timestamp changes). With this change, updating the README would happen once a year max which sounds reasonable.

See the diff here if that's not clear: https://github.com/benjamin-thomas/caqti-study/commit/aa9c672a18fa1332ab3993829d14c92d9c67f16d

**Comment by paurkedal** [comment]
> For this to work, a little change would be required (my second suggestion): reporting the stolen bikes at the beginning of the year (this is to prevent mdx continously failing because the current timestamp changes). With this change, updating the README would happen once a year max which sounds reasonable.

How about adjusting the `report_stolen` to take an optional date argument?  I think that's also closer to what one would do in a real application.

**Comment by benjamin-thomas** [comment]
> How about adjusting the `report_stolen` to take an optional date argument? I think that's also closer to what one would do in a real application.

Yes that'd be even better!

**Comment by ada2k** [comment]
benjamin_thomas: could the version you linked be merged? I don't use Github much and forgot about this PR - mdx integration is a big improvement.

**Comment by benjamin-thomas** [comment]
> benjamin_thomas: could the version you linked be merged? I don't use Github much and forgot about this PR - mdx integration is a big improvement.

Thanks for reaching out @ada2k. What I'll do is merge your PR first, then I'll apply the minor adjustments we talked about here, as a second step. Sounds like the best option :)
